### ISSUE TITLE
Help with implementation of principal stretch hyperelasticity

### ISSUE THREAD

**Issue by mrgprasad** [thelfer/tfel#419]
Hello MFront users,

I am implementing a hyperelastic model based on principal stretches (Ogden). The theoretical background for the implementation is provided in this document: 
[stretch_hyperelaasticity_impl.pdf](https://github.com/thelfer/tfel/files/12312723/stretch_hyperelaasticity_impl.pdf)

Here's my MFront implementation (MFront beginner, please pardon my obvious mistakes):
[OgdenModel.txt](https://github.com/thelfer/tfel/files/12360887/OgdenModel.txt)

I need your help in making this better:
1. I have written two functions (`m_dyad` , `m_symdy`), but I think I cannot use `@Function` together with `@DSL DefaultFiniteStrain`. What alternative do you suggest for this to work ??
- `m_dyad` computes the dyadic product of vectors
- `m_symdy` computes the symmetric dyadic product of vectors 

2. If there are other mistakes in my code (OR) unnecessary computations that can be simplified: please do point them out.

FYI: Solver is code_aster

**Comment by thelfer** [comment]
Hi @mrgprasad,

Before I dig in your implementation, did you have a look at this page: https://thelfer.github.io/tfel/web/ogden.html ?

Concerning your questions:

1. `@Function` is not a keyword of the `DefaultFiniteStrain` keyword. See ` mfront --help-keywords-list=DefaultFiniteStrain` for the list of keywords associated with this DSL. For what you need, you may either use `@Includes` or `@Members`, though `@Members`  seems better:
   ```cxx
   @Members{
   static void m_dyad(tfel::math::tvector<6u, real> mat_B, tfel::math::tvector<6u, real> mat_C) {
    tfel::math::tmatrix<6u, 6u, real>& dyad_D
    for (unsigned short k1 = 0; k1 < 6; ++k1) {
      for (unsigned short k2 = 0; k2 < 6; ++k2) {
        dyad_D(k1, k2) = mat_B(k1) * mat_C(k2);
      }
    }
  }
   ```
  However, you may want to directy use `Stensor::buildFromVectorDiadicProduct`
2. Please see if the official Ogden implementation is not already sufficient and please point out why it isn't :)

Thomas

**Comment by mrgprasad** [comment]
Hi @thelfer, thanks for the quick reply.

Here's my python script used to obtain the Ogden parameters (GitHub doesn't let me upload a file with .py extension):
[fit_hyper_ogden.py.txt](https://github.com/thelfer/tfel/files/12313255/fit_hyper_ogden.py.txt)

Here's the fitting result obtained:
![fit_Ogden](https://github.com/thelfer/tfel/assets/26977833/fdf85c8a-bbd9-4a85-a3e7-fae0109e9309)

I extended the [official Ogden implementation](https://thelfer.github.io/tfel/web/ogden.html) to order 3 and ran a simulation (uniaxial tension) with the parameters obtained from the above fitting script. The simulation results are **not comparable** to fitting results. They are orders of magnitude different.
 
I am not able to clearly identify the issue, hence I started to implement the Ogden model from scratch.

**Comment by thelfer** [comment]
@mrgprasad I used the official Ogden implementation with your coefficients (I just changed the coefficients so that everything is positive). It seems more or less close to your curve:

![image](https://github.com/thelfer/tfel/assets/13975528/bf98958f-6bd2-406e-a47c-ac85e7dfbe04)

Please find attached the implementation (which is quick and dirtry) and the MTest test.

[test.zip](https://github.com/thelfer/tfel/files/12316435/test.zip)

**Comment by mrgprasad** [comment]
@thelfer Were you able to investigate further?

Or any comments on my MFront implementation?

**Comment by thelfer** [comment]
Dear @mrgprasad Unfortunately not. I am currently on vacations before a two weeks summer school (https://mealor2.sciencesconf.org/). I'll try to investigate your issue as soon as I can, but you should not expect this to happen in the next weeks :)

**Comment by mrgprasad** [comment]
@thelfer Thank you, have a nice vacation and summer school program :) 

Update: I have updated my MFront implementation (attached in the 1st comment). I have issues with solution convergence.

**Comment by thelfer** [comment]
Dear @mrgprasad,

I took some time to analyse the behaviour of the Ogden model in the official implementation and tried to reproduce the last Figure of this page (which tries to reproduce the results of Treolar).

I finally understood that I took different conventions that the one used by most authors. In particular, you will see that I had to divide the exponents by 2 and multiply the shear modulus by 2 (Basically, I used the eigen values of the right cauchy tensor rather than the principal stretches).

With this, the official implementation, once extended for three terms (see the quick and dirty implementation attached) seems to lead to the approximately consistent results (I did not have a look at the volumetric part):

![image](https://github.com/thelfer/tfel/assets/13975528/cef1199a-a6a8-4c72-9ac4-b3310887d3f9)

However,  I could not use your values. The better would be to formally write the strain energy that you want to implement.

Thomas


[Ogden.zip](https://github.com/thelfer/tfel/files/12407552/Ogden.zip)
